File: libmuscle/python/libmuscle/test/test_checkpoint_triggers.py
import time
import pytest
from ymmsl import CheckpointRangeRule, CheckpointAtRule, Checkpoints

from libmuscle.checkpoint_triggers import (
    CombinedCheckpointTriggers, AtCheckpointTrigger, RangeCheckpointTrigger,
    TriggerManager)


def test_at_checkpoint_trigger():
    trigger = AtCheckpointTrigger([CheckpointAtRule([1.0, 3.0, 4.0, 4.5, 9.0])])

    assert trigger.next_checkpoint(0) == 1
    assert trigger.previous_checkpoint(0) is None

    assert trigger.next_checkpoint(1) == 3
    assert trigger.previous_checkpoint(1) == 1

    eps = 1e-16
    assert trigger.next_checkpoint(1 - eps) == 1
    assert trigger.previous_checkpoint(1 - eps) is None

    assert trigger.next_checkpoint(3.9) == 4
    assert trigger.previous_checkpoint(3.9) == 3

    assert trigger.next_checkpoint(4.1) == 4.5
    assert trigger.previous_checkpoint(4.1) == 4

    assert trigger.next_checkpoint(5) == 9
    assert trigger.previous_checkpoint(5) == 4.5

    assert trigger.next_checkpoint(9) is None
    assert trigger.previous_checkpoint(9) == 9

    assert trigger.next_checkpoint(11) is None
    assert trigger.previous_checkpoint(11) == 9


def test_range_checkpoint_trigger():
    range = CheckpointRangeRule(start=0.0, stop=20.0, every=1.2)
    trigger = RangeCheckpointTrigger(range)

    assert trigger.next_checkpoint(-1) == 0
    assert trigger.previous_checkpoint(-1) is None

    assert trigger.next_checkpoint(0) == pytest.approx(1.2)
    assert trigger.previous_checkpoint(0) == 0

    assert trigger.next_checkpoint(8) == pytest.approx(8.4)
    assert trigger.previous_checkpoint(8) == pytest.approx(7.2)

    assert trigger.next_checkpoint(18.2) == pytest.approx(19.2)
    assert trigger.previous_checkpoint(18.2) == pytest.approx(18)

    assert trigger.next_checkpoint(20) is None
    assert trigger.previous_checkpoint(20) == pytest.approx(19.2)


def test_range_checkpoint_trigger_default_stop():
    range = CheckpointRangeRule(start=1.0, every=1.2)
    trigger = RangeCheckpointTrigger(range)

    assert trigger.next_checkpoint(-1.) == 1
    assert trigger.previous_checkpoint(-1.) is None

    assert trigger.next_checkpoint(148148.) == pytest.approx(148148.2)
    assert trigger.previous_checkpoint(148148.) == pytest.approx(148147)

    assert trigger.next_checkpoint(148148148.) == pytest.approx(148148149)
    assert trigger.previous_checkpoint(148148148.) == pytest.approx(148148147.8)


def test_range_checkpoint_trigger_default_start():
    range = CheckpointRangeRule(every=1.2, stop=10.0)
    trigger = RangeCheckpointTrigger(range)

    assert trigger.next_checkpoint(10) is None
    assert trigger.previous_checkpoint(10) == pytest.approx(9.6)

    assert trigger.next_checkpoint(0.0) == pytest.approx(1.2)
    assert trigger.previous_checkpoint(0.0) == pytest.approx(0.0)

    assert trigger.next_checkpoint(-148148.) == pytest.approx(-148147.2)
    assert trigger.previous_checkpoint(-148148.) == pytest.approx(-148148.4)


def test_combined_checkpoint_trigger_every_at():
    rules = [CheckpointRangeRule(every=10.0), CheckpointAtRule([3.0, 7.0, 13.0, 17.0])]
    trigger = CombinedCheckpointTriggers(rules)

    assert trigger.next_checkpoint(-11.) == pytest.approx(-10)
    assert trigger.previous_checkpoint(-11) == pytest.approx(-20)

    assert trigger.next_checkpoint(0.) == pytest.approx(3)
    assert trigger.previous_checkpoint(0.) == pytest.approx(0)

    assert trigger.next_checkpoint(8.3) == pytest.approx(10)
    assert trigger.previous_checkpoint(8.3) == pytest.approx(7)

    assert trigger.next_checkpoint(14.2) == pytest.approx(17)
    assert trigger.previous_checkpoint(14.2) == pytest.approx(13)

    assert trigger.next_checkpoint(25.2) == pytest.approx(30)
    assert trigger.previous_checkpoint(25.2) == pytest.approx(20)


def test_combined_checkpoint_trigger_at_ranges():
    rules = [CheckpointAtRule([3.0, 7.0, 13.0, 17.0]),
             CheckpointRangeRule(start=0.0, every=5.0, stop=20.0),
             CheckpointRangeRule(start=20.0, every=20.0, stop=100.0)]
    trigger = CombinedCheckpointTriggers(rules)

    assert trigger.next_checkpoint(-11.) == pytest.approx(0)
    assert trigger.previous_checkpoint(-11) is None

    assert trigger.next_checkpoint(0.) == pytest.approx(3)
    assert trigger.previous_checkpoint(0.) == pytest.approx(0)

    assert trigger.next_checkpoint(8.3) == pytest.approx(10)
    assert trigger.previous_checkpoint(8.3) == pytest.approx(7)

    assert trigger.next_checkpoint(14.2) == pytest.approx(15)
    assert trigger.previous_checkpoint(14.2) == pytest.approx(13)

    assert trigger.next_checkpoint(19.3) == pytest.approx(20)
    assert trigger.previous_checkpoint(19.3) == pytest.approx(17)

    assert trigger.next_checkpoint(25.2) == pytest.approx(40)
    assert trigger.previous_checkpoint(25.2) == pytest.approx(20)

    assert trigger.next_checkpoint(95.2) == pytest.approx(100)
    assert trigger.previous_checkpoint(95.2) == pytest.approx(80)

    assert trigger.next_checkpoint(125.2) is None
    assert trigger.previous_checkpoint(125.2) == pytest.approx(100)


def test_trigger_manager_reference_time():
    monotonic_start = time.monotonic()
    ref_elapsed = 15.0
    trigger_manager = TriggerManager()
    trigger_manager.set_checkpoint_info(ref_elapsed, Checkpoints(at_end=True))
    elapsed_walltime = trigger_manager.elapsed_walltime()
    duration = time.monotonic() - monotonic_start
    assert ref_elapsed < elapsed_walltime < (ref_elapsed + duration)


def test_trigger_manager():
    ref_elapsed = 0.0
    trigger_manager = TriggerManager()
    trigger_manager.set_checkpoint_info(ref_elapsed, Checkpoints(
            at_end=True,
            wallclock_time=[CheckpointAtRule([1e-12])],
            simulation_time=[CheckpointAtRule([1.0, 3.0, 5.0])]))

    assert trigger_manager.should_save_snapshot(0.1)
    triggers = trigger_manager.get_triggers()
    assert len(triggers) == 1
    assert "wallclock_time" in triggers[0]
    trigger_manager.update_checkpoints(0.1)

    assert not trigger_manager.should_save_snapshot(0.99)

    assert trigger_manager.should_save_snapshot(3.2)
    triggers = trigger_manager.get_triggers()
    assert len(triggers) == 1
    assert "simulation_time" in triggers[0]
    trigger_manager.update_checkpoints(3.2)

    assert trigger_manager.should_save_final_snapshot(True, 7.0)
    assert len(trigger_manager.get_triggers()) > 0
    trigger_manager.update_checkpoints(7.0)

    assert not trigger_manager.should_save_snapshot(7.1)

    assert trigger_manager.should_save_final_snapshot(False, None)
    trigger_manager.update_checkpoints(7.1)


def test_no_checkpointing() -> None:
    trigger_manager = TriggerManager()
    trigger_manager.set_checkpoint_info(0.0, Checkpoints())

    assert not trigger_manager.should_save_snapshot(1)
    assert not trigger_manager.should_save_snapshot(5000)
    assert not trigger_manager.should_save_final_snapshot(False, None)

File: libmuscle/cpp/src/libmuscle/tests/test_checkpoint_triggers.cpp
#include <gtest/gtest.h>

#include <libmuscle/checkpoint_triggers.hpp>

using libmuscle::impl::TriggerManager;
using libmuscle::impl::AtCheckpointTrigger;
using libmuscle::impl::RangeCheckpointTrigger;
using libmuscle::impl::CombinedCheckpointTriggers;
using libmuscle::impl::Data;

int main(int argc, char *argv[]) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

TEST(libmuscle_checkpoint_triggers, test_at_checkpoint_trigger) {
    std::vector<double> at = {1, 3, 4, 4.5, 9};
    AtCheckpointTrigger trigger(at);

    ASSERT_DOUBLE_EQ(trigger.next_checkpoint(0).get(), 1);
    ASSERT_FALSE(trigger.previous_checkpoint(0).is_set());

    ASSERT_DOUBLE_EQ(trigger.next_checkpoint(1).get(),  3);
    ASSERT_DOUBLE_EQ(trigger.previous_checkpoint(1).get(), 1);

    double eps = 1e-16;
    ASSERT_DOUBLE_EQ(trigger.next_checkpoint(1 - eps).get(), 1);
    ASSERT_FALSE(trigger.previous_checkpoint(1 - eps).is_set());

    ASSERT_DOUBLE_EQ(trigger.next_checkpoint(3.9).get(), 4);
    ASSERT_DOUBLE_EQ(trigger.previous_checkpoint(3.9).get(), 3);

    ASSERT_DOUBLE_EQ(trigger.next_checkpoint(4.1).get(), 4.5);
    ASSERT_DOUBLE_EQ(trigger.previous_checkpoint(4.1).get(), 4);

    ASSERT_DOUBLE_EQ(trigger.next_checkpoint(5).get(), 9);
    ASSERT_DOUBLE_EQ(trigger.previous_checkpoint(5).get(), 4.5);

    ASSERT_FALSE(trigger.next_checkpoint(9).is_set());
    ASSERT_DOUBLE_EQ(trigger.previous_checkpoint(9).get(), 9);

    ASSERT_FALSE(trigger.next_checkpoint(11).is_set());
    ASSERT_DOUBLE_EQ(trigger.previous_checkpoint(11).get(), 9);
}

TEST(libmuscle_checkpoint_triggers, test_range_checkpoint_trigger) {
    auto range = Data::dict("start", 0, "stop", 20, "every", 1.2);
    RangeCheckpointTrigger trigger(range);

    ASSERT_DOUBLE_EQ(trigger.next_checkpoint(-1).get(), 0);
    ASSERT_FALSE(trigger.previous_checkpoint(-1).is_set());

    ASSERT_DOUBLE_EQ(trigger.next_checkpoint(0).get(), 1.2);
    ASSERT_DOUBLE_EQ(trigger.previous_checkpoint(0).get(), 0);

    ASSERT_DOUBLE_EQ(trigger.next_checkpoint(8).get(), 8.4);
    ASSERT_DOUBLE_EQ(trigger.previous_checkpoint(8).get(), 7.2);

    ASSERT_DOUBLE_EQ(trigger.next_checkpoint(18.2).get(), 19.2);
    ASSERT_DOUBLE_EQ(trigger.previous_checkpoint(18.2).get(), 18);

    ASSERT_FALSE(trigger.next_checkpoint(20).is_set());
    ASSERT_DOUBLE_EQ(trigger.previous_checkpoint(20).get(), 19.2);
}

TEST(libmuscle_checkpoint_triggers, test_range_checkpoint_trigger_default_stop) {
    auto range = Data::dict("start", 1, "stop", Data(), "every", 1.2);
    RangeCheckpointTrigger trigger(range);

    ASSERT_DOUBLE_EQ(trigger.next_checkpoint(-1.).get(), 1);
    ASSERT_FALSE(trigger.previous_checkpoint(-1.).is_set());

    ASSERT_DOUBLE_EQ(trigger.next_checkpoint(148148.).get(), 148148.2);
    ASSERT_DOUBLE_EQ(trigger.previous_checkpoint(148148.).get(), 148147);

    ASSERT_DOUBLE_EQ(trigger.next_checkpoint(148148148.).get(), 148148149);
    ASSERT_DOUBLE_EQ(trigger.previous_checkpoint(148148148.).get(), 148148147.8);
}

TEST(libmuscle_checkpoint_triggers, test_range_checkpoint_trigger_default_start) {
    auto range = Data::dict("start", Data(), "stop", 10, "every", 1.2);
    RangeCheckpointTrigger trigger(range);

    ASSERT_FALSE(trigger.next_checkpoint(10).is_set());
    ASSERT_DOUBLE_EQ(trigger.previous_checkpoint(10).get(), 9.6);

    ASSERT_DOUBLE_EQ(trigger.next_checkpoint(0.0).get(), 1.2);
    ASSERT_DOUBLE_EQ(trigger.previous_checkpoint(0.0).get(), 0.0);

    ASSERT_DOUBLE_EQ(trigger.next_checkpoint(-148148.).get(), -148147.2);
    ASSERT_DOUBLE_EQ(trigger.previous_checkpoint(-148148.).get(), -148148.4);
}

TEST(libmuscle_checkpoint_triggers, test_combined_checkpoint_trigger_every_at) {
    auto rules = Data::nils(2);
    rules[0] = Data::dict("start", Data(), "stop", Data(), "every", 10);
    rules[1] = Data::dict("at", Data::list(3, 7, 13, 17));
    CombinedCheckpointTriggers trigger(rules);

    ASSERT_DOUBLE_EQ(trigger.next_checkpoint(-11.).get(), -10);
    ASSERT_DOUBLE_EQ(trigger.previous_checkpoint(-11).get(), -20);

    ASSERT_DOUBLE_EQ(trigger.next_checkpoint(0.).get(), 3);
    ASSERT_DOUBLE_EQ(trigger.previous_checkpoint(0.).get(), 0);

    ASSERT_DOUBLE_EQ(trigger.next_checkpoint(8.3).get(), 10);
    ASSERT_DOUBLE_EQ(trigger.previous_checkpoint(8.3).get(), 7);

    ASSERT_DOUBLE_EQ(trigger.next_checkpoint(14.2).get(), 17);
    ASSERT_DOUBLE_EQ(trigger.previous_checkpoint(14.2).get(), 13);

    ASSERT_DOUBLE_EQ(trigger.next_checkpoint(25.2).get(), 30);
    ASSERT_DOUBLE_EQ(trigger.previous_checkpoint(25.2).get(), 20);
}

TEST(libmuscle_checkpoint_triggers, test_combined_checkpoint_trigger_at_ranges) {
    auto rules = Data::nils(3);
    rules[0] = Data::dict("at", Data::list(3, 7, 13, 17));
    rules[1] = Data::dict("start", 0, "stop", 20, "every", 5);
    rules[2] = Data::dict("start", 20, "stop", 100, "every", 20);
    CombinedCheckpointTriggers trigger(rules);

    ASSERT_DOUBLE_EQ(trigger.next_checkpoint(-11.).get(), 0);
    ASSERT_FALSE(trigger.previous_checkpoint(-11).is_set());

    ASSERT_DOUBLE_EQ(trigger.next_checkpoint(0.).get(), 3);
    ASSERT_DOUBLE_EQ(trigger.previous_checkpoint(0.).get(), 0);

    ASSERT_DOUBLE_EQ(trigger.next_checkpoint(8.3).get(), 10);
    ASSERT_DOUBLE_EQ(trigger.previous_checkpoint(8.3).get(), 7);

    ASSERT_DOUBLE_EQ(trigger.next_checkpoint(14.2).get(), 15);
    ASSERT_DOUBLE_EQ(trigger.previous_checkpoint(14.2).get(), 13);

    ASSERT_DOUBLE_EQ(trigger.next_checkpoint(19.3).get(), 20);
    ASSERT_DOUBLE_EQ(trigger.previous_checkpoint(19.3).get(), 17);

    ASSERT_DOUBLE_EQ(trigger.next_checkpoint(25.2).get(), 40);
    ASSERT_DOUBLE_EQ(trigger.previous_checkpoint(25.2).get(), 20);

    ASSERT_DOUBLE_EQ(trigger.next_checkpoint(95.2).get(), 100);
    ASSERT_DOUBLE_EQ(trigger.previous_checkpoint(95.2).get(), 80);

    ASSERT_FALSE(trigger.next_checkpoint(125.2).is_set());
    ASSERT_DOUBLE_EQ(trigger.previous_checkpoint(125.2).get(), 100);
}

TEST(libmuscle_checkpoint_triggers, test_trigger_manager_reference_time) {
    auto encoded_checkpoints = Data::dict(
        "at_end", true,
        "wallclock_time", Data::list(),
        "simulation_time", Data::list()
    );

    auto start = std::chrono::steady_clock::now();
    double ref_elapsed = 15.0;

    TriggerManager trigger_manager;
    trigger_manager.set_checkpoint_info(ref_elapsed, encoded_checkpoints);
    double elapsed_walltime = trigger_manager.elapsed_walltime();
    auto duration = std::chrono::steady_clock::now() - start;
    double duration_d = std::chrono::duration<double>(duration).count();

    ASSERT_LT(ref_elapsed, elapsed_walltime);
    ASSERT_LT(elapsed_walltime, ref_elapsed + duration_d);
}

TEST(libmuscle_checkpoint_triggers, test_trigger_manager) {
    double ref_elapsed = 0.0;
    TriggerManager trigger_manager;

    auto wallclock_time_list = Data::nils(1);
    wallclock_time_list[0] = Data::dict("at", Data::list(1e-12));
    auto simulation_time_list = Data::nils(1);
    simulation_time_list[0] = Data::dict("at", Data::list(1, 3, 5));
    auto encoded_checkpoints = Data::dict(
        "at_end", true,
        "wallclock_time", wallclock_time_list,
        "simulation_time", simulation_time_list
    );
    trigger_manager.set_checkpoint_info(ref_elapsed, encoded_checkpoints);

    ASSERT_TRUE(trigger_manager.should_save_snapshot(0.1));
    auto triggers = trigger_manager.get_triggers();
    ASSERT_EQ(triggers.size(), 1);
    ASSERT_NE(triggers[0].find("wallclock_time"), std::string::npos);
    trigger_manager.update_checkpoints(0.1);

    ASSERT_FALSE(trigger_manager.should_save_snapshot(0.99));

    ASSERT_TRUE(trigger_manager.should_save_snapshot(3.2));
    triggers = trigger_manager.get_triggers();
    ASSERT_EQ(triggers.size(), 1);
    ASSERT_NE(triggers[0].find("simulation_time"), std::string::npos);
    trigger_manager.update_checkpoints(3.2);

    ASSERT_TRUE(trigger_manager.should_save_final_snapshot(true, 7.0));
    ASSERT_GT(trigger_manager.get_triggers().size(), 0);
    trigger_manager.update_checkpoints(7.0);

    ASSERT_FALSE(trigger_manager.should_save_snapshot(7.1));

    ASSERT_TRUE(trigger_manager.should_save_final_snapshot(false, {}));
    trigger_manager.update_checkpoints(7.1);
}

TEST(libmuscle_checkpoint_triggers, test_no_checkpointing) {
    TriggerManager trigger_manager;
    auto encoded_checkpoints = Data::dict(
        "at_end", false,
        "wallclock_time", Data::list(),
        "simulation_time", Data::list()
    );
    trigger_manager.set_checkpoint_info(0.0, encoded_checkpoints);
    ASSERT_FALSE(trigger_manager.should_save_snapshot(1));
    ASSERT_FALSE(trigger_manager.should_save_snapshot(5000));
    ASSERT_FALSE(trigger_manager.should_save_final_snapshot(false, {}));
}
